python/amici/pandas.py

```diff
@@ -2,174 +2,247 @@
+    """ Compute model residuals according to lists of ReturnData and ExpData.
 
     Arguments:
-        model: Model instance
-        edata_list: list of ExpData instances with experimental data
-        rdata_list: list of ReturnData instances corresponding to ExpData
+        model: Model instance.
+        edata_list: list of ExpData instances with experimental data.
+        rdata_list: list of ReturnData instances corresponding to ExpData.
+        by_id: bool, optional (default = False)
+            If True, ids are used as identifiers, otherwise the possibly more
+            descriptive names.
 
     Returns:
-        pandas DataFrame with conditions and observables
+        pandas DataFrame with conditions and observables.
 
     Raises:
 
     """
+    if isinstance(edata_list, amici.amici.ExpData):
+        edata_list = [edata_list]
+    if isinstance(rdata_list, amici.amici.ReturnData):
+        rdata_list = [rdata_list]
+
+    # get conditions and state column names by name or id
+    cols = _get_state_cols(model, by_id=by_id)
 
-    cols = _get_state_cols(model)
+    # initialize dataframe with columns
     df_rdata = pd.DataFrame(columns=cols)
 
+    # append states
     for edata, rdata in zip(edata_list, rdata_list):
         for i_time, timepoint in enumerate(rdata['t']):
             datadict = {
                 'time': timepoint,
             }
 
-            for i_state, state in enumerate(_get_names_or_ids(model, 'State')):
+            # append states
+            for i_state, state in enumerate(
+                    _get_names_or_ids(model, 'State', by_id=by_id)):
                 datadict[state] = rdata['x'][i_time, i_state]
 
-            _fill_conditions_dict(datadict, model, edata)
+            # use data to fill condition columns
+            _fill_conditions_dict(datadict, model, edata, by_id=by_id)
 
+            # append to dataframe
             df_rdata.loc[len(df_rdata)] = datadict
+
     return df_rdata
 
 
-def getResidualsAsDataFrame(model, edata_list, rdata_list):
-    """ Convert a list of ExpData to pandas DataFrame
+def getResidualsAsDataFrame(model, edata_list, rdata_list, by_id=False):
+    """ Convert a list of ExpData to pandas DataFrame.
 
     Arguments:
-        model: Model instance
-        edata_list: list of ExpData instances with experimental data
-        rdata_list: list of ReturnData instances corresponding to ExpData
+        model: Model instance.
+        edata_list: list of ExpData instances with experimental data.
+        rdata_list: list of ReturnData instances corresponding to ExpData.
+        by_id: bool, optional (default = False)
+            If True, ids are used as identifiers, otherwise the possibly more
+            descriptive names.
 
     Returns:
-        pandas DataFrame with conditions and observables
+        pandas DataFrame with conditions and observables.
 
     Raises:
 
     """
+    if isinstance(edata_list, amici.amici.ExpData):
```

Comment:
why not `amici.ExpData`? could also add `amici.ExpDataPtr` here (or whatever the swig export of the smart pointer is called)

Reply:
ah, I thought the class instance has to be called as ``amici.amici.ExpData```. Just ``amici.ExpData`` had not worked in another project some weeks ago.

Reply:
should work now, if not, please let me know.

Reply:
Doesn't seem to work: https://travis-ci.org/ICB-DCM/AMICI/jobs/496114885#L7769

Reply:
Ah right, sorry! `amici.ExpData` only exposes the constructor, not the class itself.
lets do 
```suggestion
    if isinstance(edata_list, (amici.amici.ExpData, amici.amici.ExpDataPtr)):
```
then.